Scripts/Modules/SnappingModule/SnappingModule.cs

```diff
@@ -511,65 +526,201 @@ public bool DirectSnap(Transform rayOrigin, Transform transform, ref Vector3 pos
+					if (pivotSnappingEnabled)
+					{
+						var extra = k_RayExtra * viewerScale;
+						raycastDistance += extra;
+						boundsRay.origin -= directionVector * extra;
+					}
+					else
+					{
+						boundsRay.origin += targetRotation * boundsCenter;
+					}
+
+					if (TryBreakSnap(ref position, ref rotation, targetPosition, targetRotation, state, boundsBreakDist))
+						return true;
+
+					if (surfaceSnappingEnabled && SnapToSurface(boundsRay, ref position, ref rotation, state, offset, targetRotation, rotationOffset, upVector, raycastDistance))
+						return true;
+
+					if (groundSnappingEnabled && SnapToGround(boundsRay, ref position, ref rotation, targetPosition, targetRotation, state, raycastDistance, offset))
+						return true;
+
+					return true;
+				}
+
+				if (TryBreakSnap(ref position, ref rotation, targetPosition, targetRotation, state, breakDistance))
 					return true;
 			}
 
-			if (TryBreakSurfaceSnap(ref position, ref rotation, targetPosition, targetRotation, state, breakDistance))
+			for (var i = 0; i < k_Directions.Length; i++)
+			{
+				var direction = k_Directions[i];
+				var upVector = targetRotation * direction.upVector;
+				var directionVector = direction.direction;
+				var rotationOffset = direction.rotationOffset;
+				var offset = Vector3.zero;
+				if (!pivotSnappingEnabled)
+				{
+					var projectedExtents = Vector3.Project(boundsExtents, directionVector);
+					offset = -boundsCenter;
+					if (i > 2)
+						offset -= projectedExtents;
+					else
+						offset += projectedExtents;
+
+					offset = rotation * offset;
+				}
+
+				var raycastDistance = breakDistance * 2;
+				directionVector = targetRotation * directionVector;
+				var boundsRay = new Ray(targetPosition - offset - directionVector * breakDistance, directionVector);
+
+				if (surfaceSnappingEnabled && SnapToSurface(boundsRay, ref position, ref rotation, state, offset, targetRotation, rotationOffset, upVector, raycastDistance, k_MaxRayDot))
+				{
+					state.directionIndex = i;
+					return true;
+				}
+
+				if (groundSnappingEnabled && SnapToGround(boundsRay, ref position, ref rotation, targetPosition, targetRotation, state, raycastDistance, offset))
+				{
+					state.directionIndex = i;
+					return true;
+				}
+			}
+
+			if (TryBreakSnap(ref position, ref rotation, targetPosition, targetRotation, state, breakDistance))
+				return true;
+
+			if (TryBreakSnap(ref position, ref rotation, state.snappingPosition, targetRotation, state, breakDistance))
 				return true;
 
 			return false;
 		}
 
-		static bool TryBreakSurfaceSnap(ref Vector3 position, ref Quaternion rotation, Vector3 targetPosition, Quaternion targetRotation, SnappingState state, float breakDistance)
+		static bool TryBreakSnap(ref Vector3 position, ref Quaternion rotation, Vector3 targetPosition, Quaternion targetRotation, SnappingState state, float breakDistance)
 		{
-			if (state.surfaceSnapping)
+			if (state.snapping)
```

Comment:
Would really fast motion that stayed within snapping range possible trigger a snap break?